### PR TITLE
CLDR-16185 Add era codes to supplementalData

### PR DESCRIPTION
**Pull request by sffc**
CLDR-16185

- [ ] This PR completes the ticket.

ALLOW_MANY_COMMITS=true

@Manishearth @yumaoka @FrankYFTang

### CODE REVIEW

**Review by Manishearth — DISMISSED** [review]

File: common/supplemental/supplementalData.xml
Comment (FrankYFTang) [inline common/supplemental/supplementalData.xml:4731]: the type in this calendar is "ethiopic-amete-alem" should the code in here be also "ethiopic-amete-alem" and the aliases be "ethioaa mundi" instead?

**Review by FrankYFTang — COMMENTED** [review]
about consistency between code of era and type of calendar

File: common/supplemental/supplementalData.xml
Comment (FrankYFTang) [reply common/supplemental/supplementalData.xml:4731]: I understand this now.

**Comment by FrankYFTang** [comment]
We may also need to amend https://www.unicode.org/reports/tr35/tr35-dates.html#Calendar_Data

**Comment by FrankYFTang** [comment]
Should we also change https://github.com/unicode-org/cldr/blob/main/docs/ldml/tr35-dates.md#Calendar_Data in this PR?

File: common/supplemental/supplementalData.xml
Comment (sffc) [inline common/supplemental/supplementalData.xml:4652]: @justingrant suggests removing the aliases here. I agree.

Comment (macchiati) [reply common/supplemental/supplementalData.xml:4652]: We're also discussed a different approach for the codes.

**Review by macchiati — CHANGES_REQUESTED** [review]

File: common/supplemental/supplementalData.xml
Comment (sffc) [reply common/supplemental/supplementalData.xml:4652]: Yes. For context, the suggestion was to use ordinals in insertion order with post-meiji elision, resulting in era codes such as `showa-001` for "the first Showa era in the period before Meiji". I floated the idea by @Manishearth, who reminded me of his position from when we previously discussed this last year, which is that `showa-001` doesn't tell you anything about the era: was it in the first or second millennium? Before or after the Edo Period?

The counter-proposal is to use `showa-131x`. I'm currently waiting on feedback from @yumaoka.

File: common/dtd/ldmlSupplemental.dtd
Comment (macchiati) [inline common/dtd/ldmlSupplemental.dtd]: Don't remove the @VALUE. That marks that the era code is not metadata and not distinguishing, and instead is additional data about the type. Just like the start/end data

File: common/supplemental/supplementalData.xml
Comment (macchiati) [inline common/supplemental/supplementalData.xml:4414]: Question; wouldn't the japanese-inverse end on that date (or actually the day before)?

**Review by macchiati — CHANGES_REQUESTED** [review]

File: common/dtd/ldmlSupplemental.dtd
Comment (sffc) [reply common/dtd/ldmlSupplemental.dtd]: I tried, but for some reason if I have the `@VALUE`, I get the following error. I DON'T get this error if I remove it.

> Error: (TestDtdData.java:313) : isDistinguished supplementalData: elementName.equals("era") && attribute.equals("code"): expected true, got false

**Review by Manishearth — APPROVED** [review]

File: common/dtd/ldmlSupplemental.dtd
Comment (macchiati) [reply common/dtd/ldmlSupplemental.dtd]: I'll have to look it over tomorrow morning

File: common/supplemental/supplementalData.xml
Comment (macchiati) [reply common/supplemental/supplementalData.xml]: ?

**Review by macchiati — CHANGES_REQUESTED** [review]
It is clear that eras with different start/end dates must have different codes. For those that have identical start/end dates, **but different fundamental semantics**, we also need different era codes. However, the following don't make sense to me:

```
<era type="0" start="622-7-15" code="islamic" aliases="ah"/>
<era type="0" start="622-7-15" code="islamic-rgsa" aliases="ah"/>
<era type="0" start="622-7-15" code="islamic-tbla" aliases="ah"/>
<era type="0" start="622-7-15" code="islamic-umalqura" aliases="ah"/>

```
I believe these should all be:
<era type="0" start="622-7-15" code="islamic" aliases="ah"/>

**Review by macchiati — CHANGES_REQUESTED** [review]
(see other comment)

File: common/supplemental/supplementalData.xml
Comment (sffc) [reply common/supplemental/supplementalData.xml]: Fixed.

File: common/dtd/ldmlSupplemental.dtd
Comment (sffc) [reply common/dtd/ldmlSupplemental.dtd]: Thanks @macchiati for your investigation! It is working now

**Comment by justingrant** [comment]
> eras with different start/end dates must have different codes. For those that have identical start/end dates, but different fundamental semantics, we also need different era codes.

Aren't era names scoped to each calendar?  If not&mdash;meaning we're expecting any era to be used in any calendar&mdash;then it seems like we're adding an unnecessary future compatibility risk.  What is the benefit of allowing era codes to be free-standing entities, as opposed to being scoped to each calendar like AFAICT real-world eras are?

File: common/supplemental/supplementalData.xml
Comment (sffc) [reply common/supplemental/supplementalData.xml]: The codes for pre-Meiji eras were removed in order to allow for more time to investigate the quality of the data.

**Comment by sffc** [comment]
I still need to write up the spec, but the understanding is that era codes are globally unique, and era aliases are scoped to the calendar. This allows for all the flavors of Islamic to use an "ah" era alias, as well as for Gregorian, Julian, and Juliogregorian to all use "ad" and "bc".

@macchiati I'm not sure what else is missing? The four Islamic eras have different codes and shared aliases.

**Comment by sffc** [comment]
@macchiati Are you suggesting that the four Islamic eras all share the same era code? I gave them different codes primarily because they have different behavior.

**Comment by macchiati** [comment]
All of those calendars are closely related variants of the islamic
calendar, and share the fundamental semantics and start/end dates. I don't
see the value in having separate codes; only user confusion.

**Comment by sffc** [comment]
@macchiati My [proposal](https://github.com/tc39/proposal-intl-era-monthcode/issues/9) suggests that an era code defines two things:

> 1. An epoch day
> 2. A formula for converting from the (eraYear, monthCode, day) triple to a number of days relative to that epoch

The reasons for constraint 2 are:

1. Allows for eras with the same epoch day to coexist, like `gregory` and `julian` may need to do in Juliogregory.
2. Allows for unique identifiers during date conversion.

Example to illustrate point 2: suppose that a user keeps dates in the Umm al-Qura for bookkeeping purposes, but wants to render them in the Tabular Islamic calendar based on moon sightings. Having separate codes for the two eras allows for the following code to be written: `{ calendar: "islamic-tbla", era: "islamic-umalqura", eraYear: 1444, monthCode: "M07", day: 12 }`. This is _similar_ to how a date could be expressed in the `ethioaa` era but then converted at the call site to the `ethiopic` one.

In addition to this, each calendar has an era of the same name as the calendar so that it can be the anchor era for the relativeYear / arithmetic year.

**Comment by macchiati** [comment]
#2 seems very hacky to me, convoluting the era names just for a slightly
more compact form in some cases than explicitly including the calendar:

The full description appears to be that you want to convert from
{calendar:"islamic-umalqura" era: "islamic", eraYear: 1444, monthCode:
"M07", day: 12} to calendar: "islamic-tbla"
Just like I might want to convert from
{calendar:"japanese" era: "showa", eraYear: 3, monthCode: "M07", day: 12}
 to calendar: "islamic-tbla"

Because the eras are often distinct for calendars, you are forcing all of
them to be fully distinct so you can save a parameter. But although I find
it screwy, I won't stand in the way.

**Comment by justingrant** [comment]
> suppose that a user keeps dates in the Umm al-Qura for bookkeeping purposes, but wants to render them in the Tabular Islamic calendar based on moon sightings. Having separate codes for the two eras allows for the following code to be written: `{ calendar: "islamic-tbla", era: "islamic-umalqura", eraYear: 1444, monthCode: "M07", day: 12 }`.

I'm not sure I understand this example. My understanding was that the way to move dates between calendars was to use `withCalendar()`, like this:

```js
date = Temporal.PlainDate.from({ calendar: "islamic-umalqura", year: 1444, monthCode: "M07", day: 12 });
tabluarDate = date.withCalendar('islamic-tbla');

date = Temporal.PlainDate.from({ calendar: "japanese", eraYear: 10, era: 'meiji', monthCode: "M07", day: 12 });
gregorianDate = date.witihCalendar('gregory');
```

I think the core issue here is that it's confusing to apply an era from one calendar to a different calendar where that era isn't used.  This confuses the concepts of calendars and eras. These concepts are _already_ confusing on their own. Mixing them up makes it worse. 

I think this is even more confusing in single-era calendars like Islamic calendars, Chinese, Buddhist, etc. where eras aren't generally used in normal life.  In those cases, the era and the calendar are essentially the same thing, which makes it confusing when they're mixed. And even more confusing if the (single) era has the same name as the calendar. 

That said, it's an important question how to define and name eras in calendars that _do_ use eras, like:
* Japanese which (unlike all other CLDR calendars) there's no obvious "primary era" to pick, so we want to use AD as the anchor era which determines `year`
* Ethiopian calendars that are only differentiated by eras
* Gregorian and Julian-Gregorian, which will likely share same-named eras.

Some calendars like the ones above may share same-named and/or same-semantics eras with other calendars. But these special cases don't require a generalized ability to use any era across to any calendar.  I would argue that this kind of many-to-many mapping between calendars and eras makes both concepts harder to understand.

Overall, I think that eras will be easier to explain to programmers if:
* Eras are scoped to each calendar, so there's no expectation they can be used in calendars where they don't belong.
* Different calendars may have eras with the same name, without this being a requirement for all eras to be shared across all calendars
* Each era's canonical name (the name that shows up in the debugger) is a name like `bc` or `bce` that's familiar to users of that calendar, instead of making up names like `gregory-inverse` which require a trip to the docs for programmers to understand and that doesn't improve discoverability because its name simply matches the calendar's name.

Note that my concerns are about the public API, not CLDR data. If there's some reason for the two to differ, that's OK with me, as long as it doesn't make it harder to have an understandable public API.

File: common/supplemental/supplementalData.xml
Comment (FrankYFTang) [inline common/supplemental/supplementalData.xml:4741]: I suggest you keep it simple and remove aliases="broc" . It is not a known code and there are no need for such alias.

Comment (sffc) [reply common/supplemental/supplementalData.xml:4741]: When writing this PR, I adopted all narrow abbreviations in the root and english locales as aliases.

Comment (FrankYFTang) [reply common/supplemental/supplementalData.xml:4741]: but four letter "broc" is nowhere to be found in CLDR files anywhere now. How is that related to "all narrow abbreviations in the root and english locales" ?  Which file has that?

I did a 
```
find|xargs egrep -s   broc
```
but none of them are broc as a token and related to a era.

Comment (Manishearth) [reply common/supplemental/supplementalData.xml:4741]: Yes, but I'm not 100% sure if that is attested. As Frank found there is definitely 民國前 in the wild, but I'm not sure about "B. ROC." or similar.

A thing I've repeatedly come across when it comes to CLDR non-gregorian calendar data is that a lot of the symbols aren't necessarily attested and may just be made up.

Comment (sffc) [reply common/supplemental/supplementalData.xml:4741]: https://github.com/unicode-org/cldr/blob/abdaabaef38362e3cef2b20b4938c0a5f16cd902/common/main/en.xml#L2825

```xml
			<calendar type="roc">
				<eras>
					<eraAbbr>
						<era type="0">B.R.O.C.</era>
						<era type="1">Minguo</era>
					</eraAbbr>
				</eras>
			</calendar>
```

Comment (FrankYFTang) [reply common/supplemental/supplementalData.xml:4741]: to be clear. I am not asking you to change to a different alias for "roc-inverse" I am just saying there are no need for any alias for it. Keep it empty until someone have a strong case to lobby to add one (I do NOT).

Comment (sffc) [reply common/supplemental/supplementalData.xml]: OK, removed.

Comment (FrankYFTang) [reply common/supplemental/supplementalData.xml]: oh I see. I suggest you to use
"before-roc" instead. At least that is already used in several  API

Java https://docs.oracle.com/javase/8/docs/api/java/time/chrono/MinguoEra.html
Android https://developer.android.com/reference/java/time/chrono/MinguoEra#BEFORE_ROC
Joda Time https://cljdoc.org/d/clojure-interop/java.time/1.0.5/api/jdk.time.chrono.MinguoEra#BEFORE_ROC
HarmonyOS https://developer.harmonyos.com/en/docs/documentation/doc-references/minguoera-0000001054518588#ZH-CN_TOPIC_0000001054518588__BEFORE_ROC
Three Ten https://www.threeten.org/threetenbp/apidocs/org/threeten/bp/chrono/MinguoEra.html#BEFORE_ROC

Comment (FrankYFTang) [reply common/supplemental/supplementalData.xml]: I am also fine if you do not add "before-roc"

Comment (sffc) [reply common/supplemental/supplementalData.xml]: Added "before-roc" as an alias.

**Review by macchiati — APPROVED** [review]

**Comment by macchiati** [comment]
@sffc if you don't have merge rights, let me know.

**Review by FrankYFTang — APPROVED** [review]
Thanks for working on this!!!

**Comment by FrankYFTang** [comment]
I think he need to squashed it first.

**Review by FrankYFTang — APPROVED** [review]

**Comment by robertbastian** [comment]
I agree with Mark's and Justin's sentiment that calendars should be able to share era names. In particular

* Why do the islamic calendars not all call their era `islamic`? They all allow calling it `ah`, so the calendarless validation that Shane seems to want to do does not always work.
* Why does the Japanese calendar call its Gregorian eras `japanese-inverse` and `japanese` instead of `gregory-inverse` and `gregory`? "The Japanese era" is not a thing that exists, these names seem to also have been chosen for the purpose of guessing calendar from era.

Both of these instances break the property that an identifier is either an era name, or an alias to another era name: in the Islamic case because `ah` is an alias to four different era names, in the Japanese case because `gregory` is an alias to `japanese` in the Japanese calendar, but an era in the Gregorian calendar. This makes it impossible to normalise an era name without knowing the calendar to normalize for.

I don't understand why this ticket introduces the requirement that era codes need to be unique across calendars. The previous era numbers certainly weren't.

**Comment by sffc** [comment]
With the landed CLDR proposal, calendars _can_ share era names. It's only when you "canonicalize" an era that it produces a code that is globally unique. I stated my original motivation in https://github.com/unicode-org/cldr/pull/2665#issuecomment-1414577550, but I acknowledge the points in this thread and agree that there's likely more iteration that could be done. One suggestion I support is being proposed by @ptomato in https://github.com/tc39/proposal-intl-era-monthcode/pull/20

**Pull request merged**
